### PR TITLE
Feature/pass args to webpack

### PR DESCRIPTION
**Pull request by JakeTheCorn**
This PR is intended to simplify the build by...

1.)  Removing 100+ LOC from webpack.config.js
2.)  Organizing the build process into stages.

This also improves the way we are passing argument overrides.  
   -- the main logic to this resides in config/webpack/argumentOverride.js

Mostly, the increase in modularity alone makes the build more SRP and easy to debug.

I'm sure there's room for improvement and look forward to hearing back from @BWortman .

Thanks!

--Jake

### CODE REVIEW

File: config/webpack/argumentOverride.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  for (let argKey in args) {
+    for (let optionKey in defaultOptions) {
+      key = optionKey;
+      if (argKey.toLowerCase() === optionKey.toLowerCase()) {
+        defaultOptions[key] = args[argKey];
+      }
+    }
+  }
+  if (defaultOptions.isProd && !args.targetUrl) {
+    defaultOptions.targetUrl = 'https://autorenter-nodeexpress-api.herokuapp.com/api/';
```
Comment (BWortman) [inline config/webpack/argumentOverride.js]: This URL also appears in the defaultOptions.js file.

File: config/webpack/defaultOptions.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+const buildDefaultOptions = (dir) => {
+  const isProd = process.env.__PROD__ == 'true' || process.env.isProd;
+  const targetUrl = isProd ? "https://autorenter-nodeexpress-api.herokuapp.com/api/" : "http://127.0.0.1:3000/api/";
```
Comment (BWortman) [inline config/webpack/defaultOptions.js]: This URL also appears in the argumentOverride file. Any way to make it DRY?

I think we should take the env.js approach, described here: https://www.jvandemo.com/how-to-configure-your-angularjs-application-using-environment-variables/. Let's the the target API url out of the code and build  configuration.

**Comment by JakeTheCorn** [comment]
@BWortman 

are we only using `npm start` in production since we have a different command for the dev server?

**Comment by BWortman** [comment]
I believe so, yes. It's been more or less a black box to me (which I've left up to others) up to this point.


In any case, I think we should go with the env.js approach for this particular problem.

________________________________

**Comment by JakeTheCorn** [comment]
@BWortman 

Okay, cool. 

Unless I'm mistaken, the main goal is to hide the connection strings and whatnot from the public code in order to improve scaling, privacy, efficiency.....etc. 

If thats the case, I've almost got a .dotenv solution to push up.

**Comment by JakeTheCorn** [comment]
@BWortman 

Does Fusion have a private repo to secure/centralise application keys, etc?  or how do we do this

**Comment by BWortman** [comment]
This addresses that concern: https://www.jvandemo.com/how-to-configure-your-angularjs-application-using-environment-variables/ (near the bottom).


At this point we only have the API URL to add to the env.js file. The API is public, and it doesn't require an API token. So it's okay putting it in our version control.


Someday it would be nice to separate deployment from code (McConnell's good idea I mentioned yesterday), but that's for another time 😉



________________________________

**Comment by JakeTheCorn** [comment]
there were issues following his example verbatim.  namely that referencing the env file in index.html doesn't work in development because of the way webpack dev server reads from in memory and not by reference.

if you run `npm run build` which is the dependent command for npm start you can then search the resulting app****.js file under dist and you will find the find the proper string for api url read only from our dot env file which achieves the same goal.

works in dev.

environment variables are now tucked safely away, redundant code is removed.

Let's open an issue to separate deployment from code?

File: .gitignore [hunk 3]
```diff
@@ -1,5 +1,8 @@
 # See https://help.github.com/articles/ignoring-files/ for more about ignoring files.
 
+# Environment Variables
+.env
```
Comment (BWortman) [inline .gitignore]: Good work, @JakeTheCorn. Can you add an .env-example file to the repo? Actually, I wouldn't be averse to adding an actual .env, with comments that it should never include any sensitive info.

If you only add the .env-example, be sure to explain it in the README.

**Comment by BWortman** [comment]
As a separate but related issue, this is in the karma.conf.js file:

var isProd = process.env.__PROD__ == 'true' || process.env.isProd || args.prod;
var targetUrl = isProd ? "https://autorenter-nodeexpress-api.herokuapp.com/api/" : "http://127.0.0.1:3000/api/";
var apiUrl = (process.env.__APIURL__ == undefined) ? targetUrl : process.env.__APIURL__;

File: config/defaultOptions.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
+require('dotenv').config();
+const args = require('yargs').argv;
+
+const buildDefaultOptions = (dir) => {
+  const isProd = args.prod;
+  const targetUrl = isProd ? process.env.PRODUCTION_URL : process.env.DEVELOPMENT_URL;
```
Comment (BWortman) [inline config/defaultOptions.js]: Let's just have a single target url var for the api, not separate env vars for prod and dev. The appropriate url will be configured based on the environment.

**Comment by BWortman** [comment]
Added @camoore1979, in case he has any thoughts.

Thanks, @JakeTheCorn.

**Comment by JakeTheCorn** [comment]
@BWortman 

I saw the build issue locally....

So, bigger problem at hand is >>

"Why didn't our CI catch that?"

Should we add cleaning scripts to the CI process?

**Pull request merged**